src/impl.cpp

```diff
@@ -318,75 +215,105 @@ void Manifold::Impl::InitializeOriginal(bool keepFaceID) {
-          const int prop1 =
-              meshRelation_
-                  .triProperties[pairFace][jointNum == 2 ? 0 : jointNum + 1];
-          bool propEqual = true;
-          for (size_t p = 0; p < numProp; ++p) {
-            if (meshRelation_.properties[numProp * prop0 + p] !=
-                meshRelation_.properties[numProp * prop1 + p]) {
-              propEqual = false;
-              break;
-            }
-          }
-          if (propEqual) {
-            vert2vert[edgeIdx] = std::make_pair(prop0, prop1);
-          }
-        });
-    DedupePropVerts(meshRelation_.triProperties, vert2vert);
-  }
+  const int numTri = NumTri();
+  struct TriPriority {
+    double area2;
+    int tri;
+  };
+  Vec<TriPriority> triPriority(numTri);
+  for_each_n(autoPolicy(numTri), countAt(0), numTri,
+             [&triPriority, this](int tri) {
+               meshRelation_.triRef[tri].faceID = -1;
+               const vec3 v = vertPos_[halfedge_[3 * tri].startVert];
+               triPriority[tri] = {
+                   length2(cross(vertPos_[halfedge_[3 * tri].endVert] - v,
+                                 vertPos_[halfedge_[3 * tri + 1].endVert] - v)),
+                   tri};
+             });
 
-  for_each_n(autoPolicy(halfedge_.size(), 1e4), countAt(0), halfedge_.size(),
-             CoplanarEdge({face2face, triArea, halfedge_, vertPos_,
-                           meshRelation_.triRef, meshRelation_.triProperties,
-                           meshRelation_.numProp, epsilon_, tolerance_}));
-
-  std::vector<int> components;
-  const int numComponent = GetLabels(components, face2face, NumTri());
-
-  Vec<int> comp2tri(numComponent, -1);
-  for (size_t tri = 0; tri < NumTri(); ++tri) {
-    const int comp = components[tri];
-    const int current = comp2tri[comp];
-    if (current < 0 || triArea[tri] > triArea[current]) {
-      comp2tri[comp] = tri;
-      triArea[comp] = triArea[tri];
+  stable_sort(triPriority.begin(), triPriority.end(),
+              [](auto a, auto b) { return a.area2 > b.area2; });
+
+  for (const auto tp : triPriority) {
+    if (meshRelation_.triRef[tp.tri].faceID >= 0) continue;
+
+    meshRelation_.triRef[tp.tri].faceID = tp.tri;
+    const vec3 base = vertPos_[halfedge_[3 * tp.tri].startVert];
+    const vec3 normal = faceNormal_[tp.tri];
+    std::deque<int> interiorHalfedges = {3 * tp.tri, 3 * tp.tri + 1,
+                                         3 * tp.tri + 2};
+    while (!interiorHalfedges.empty()) {
+      const int h =
+          NextHalfedge(halfedge_[interiorHalfedges.back()].pairedHalfedge);
+      interiorHalfedges.pop_back();
+      if (meshRelation_.triRef[h / 3].faceID >= 0) continue;
+
+      const vec3 v = vertPos_[halfedge_[h].endVert];
+      if (abs(dot(v - base, normal)) < tolerance_) {
+        meshRelation_.triRef[h / 3].faceID = tp.tri;
+        if (!interiorHalfedges.empty() &&
+            h == halfedge_[interiorHalfedges.back()].pairedHalfedge) {
```

Comment:
Considering `h = NextHalfedge(halfedge_[interiorHalfedges.back()].pairedHalfedge)`, will this condition ever hold?

Reply:
No, because the `back` got popped off in line 248.